### PR TITLE
Hide ad slots if targeting consent is denied

### PR DESCRIPTION
**Pull request by Jakeii**
## What does this change?
Hides ad slots when consent has been denied.

**Edit:**
The scope of this issue has expanded somewhat due to the fact that if the ad free cookie can now be set for different reasons, we need to keep track of them so that the ad free cookie is not removed if one reason goes away but not another e.g. a user is a digital subscriber who has opted out, if their subscription stops or they opt back in, they still need to be ad free.

Me & @zekehuntergreen chose to do this using a localStorage object, that is a map of the reasons for ad free and their expiry, another cookie would have also made sense but would have required getting permission and then messing with the fastly cache to split it using it.

We also thought it best to refactor and move all manipulation of the ad free cookie to a separate file so that it would not be changed without updating the localStorage reason.

We've tested all the combinations of 
- opt in / out
- logged in subscriber / logged in non-subscriber / logged out state
- the various transitions between them, logging in/out and opting in/out while a subscriber or not
- simulating expiry of them
-  going to other consent regions via VPN

Everything works as expected

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

## Screenshots

| Before      | After      |
|-------------|------------|
| Before | after |

[before]: https://example.com/before.png
[after]: https://example.com/after.png

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [x] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [x] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Does this change update the version of CAPI we're using?

- [x] No, all the existing database files are just fine
- [ ] Yes, and I have [re-run all the tests locally and checked in all the updated data/database/xyz files](https://github.com/guardian/frontend/blob/main/docs/03-dev-howtos/15-updating-test-database.md)

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [x] Locally
- [x] On CODE (optional)

### CODE REVIEW

**Comment by OllysCoding** [comment]
If you loaded this page with consent already denied, would this then cause CLS, or are the ad slots just not added into the page in stat situation?

**Comment by Jakeii** [comment]
@OllysCoding It does cause CLS, I spoke to @arelra about removing during SSR, but (sorry can't remember exactly) he mentioned that the way the cache is split would make it difficult to do.

**Comment by oliverlloyd** [comment]
> @OllysCoding It does cause CLS, I spoke to @arelra about removing during SSR, but (sorry can't remember exactly) he mentioned that the way the cache is split would make it difficult to do.

I think in this case it would be preferable (both for the reader and for our CLS scores) to keep the space open on the page?

**Comment by Jakeii** [comment]
We have had a lot of complaints about the gap (at the top specifically), perhaps a topic for next comdev/dotcom meetup?

**Comment by oliverlloyd** [comment]
> If the user rejects consent could we set `adFree`?
> 
> We have `adFree` server side, and in that case I believe we already choose not to SSR the slots.
> 
> As an option would that be acceptable @OllysCoding @oliverlloyd ?

So if I reject consent then my subsequent requests include the `adFree` status so on the server we don't even render the ad slots? That sounds ideal from my point fo view.

File: static/src/javascripts/projects/commercial/modules/maybe-remove-ad-slot.ts [hunk 1]
```diff
@@ -46,7 +53,9 @@ const adFreeSlotRemove = once(() => {
 			bodyEl.classList.remove('has-active-pageskin');
 		}
 
+		$$('.top-banner-ad-container').remove();
```
[the remark this answers is not in the document]
Comment (Jakeii) [reply static/src/javascripts/projects/commercial/modules/maybe-remove-ad-slot.ts:44]: When consent is initially denied the container needs to be removed

[the remark this answers is not in the document]
Comment (zekehuntergreen) [reply static/src/javascripts/projects/commercial/modules/maybe-remove-ad-slot.ts:44]: Good call, let's keep this 👍

File: static/src/javascripts/projects/commercial/modules/manage-ad-free-cookie-on-consent-change.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * If consent changes so that targeted advertising is disabled in tcfv2 regions,
+ * remove ad slots and set adFree cookie so the server won't render them on subsequent page loads
+ * Otherwise if in tcfv2 but consent to targeting is later allowed, removes the adFree cookie
+ */
+const manageAdFreeCookieOnConsentChange = once((): Promise<void> => {
+	onConsentChange((consent) => {
+		if (consent.tcfv2) {
+			if (!consent.canTarget) {
+				setAdFreeCookie(AdFreeCookieReasons.ConsentOptOut);
```
Comment (arelra) [inline static/src/javascripts/projects/commercial/modules/manage-ad-free-cookie-on-consent-change.ts:19]: Is it worth adding a unit test for this?

I think you could copy quite a bit of `initial-consent-state.spec.ts`

File: static/src/javascripts/lib/manage-ad-free-cookie.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+			return true;
+		},
+	);
+
+	if (allExpired) {
+		removeCookie({ name: AD_FREE_USER_COOKIE });
+	}
+};
+
+export {
```
Comment (arelra) [inline static/src/javascripts/lib/manage-ad-free-cookie.ts:111]: Is there any logic here that could benefit from a unit test? Only if you think its worth it!

**Review by arelra — COMMENTED** [review]
Great work on this! Just some minor comments.

File: static/src/javascripts/projects/commercial/modules/manage-ad-free-cookie-on-consent-change.ts (hunk 2 above)
Comment (Jakeii) [reply static/src/javascripts/projects/commercial/modules/manage-ad-free-cookie-on-consent-change.ts:19]: good point!

**Review by arelra — APPROVED** [review]
Great work and persistence on this! 🎉

**Pull request merged**